### DISCUSSION TITLE
A guide on disabling/enabling lsp-mode features - LSP Mode - LSP support for Emacs

### DISCUSSION THREAD

**Discussion by giscus** [emacs-lsp/lsp-mode#3793] [Giscus]
# A guide on disabling/enabling lsp-mode features - LSP Mode - LSP support for Emacs

Language Server Protocol Support for Emacs

https://emacs-lsp.github.io/lsp-mode/tutorials/how-to-turn-off/

**Comment by uliw** [comment]
great visual guide, but regardless what I set in my init file, the values are always reset to their default values (setting them after loading lsp-mode works fine). I spefically have the following two lines in init.el
```
(setq lsp-ui-sideline-show-code-actions t)
(setq lsp-ui-sideline-show-diagnostics nil)
```
but to no avail
Is there something obvious I am missing here?

**Reply by ericdallo** [reply]
Do you use any emacs framework that may override those settings?

**Comment by uliw** [comment]
upon some digging, it appears that this is related to some other package in my startup file. I'll report back

**Reply by uliw** [reply]
after removing everything, and then adding it back, it now works as expected. I was not able to pin down any specific cause.

**Comment by shaze** [comment]
Thanks for this guide, but I haven't been able to disable the lsp-ui-doc feature -- which is driving me nuts (I don't want to be told what a u8 is ever 2s)

This is what I have in my init.el -- I've tried a whole range of solutions but about to abandon this all. But if you have any ideas that would be great.

```
(custom-set-variables
 ;; custom-set-variables was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 '(gud-gdb-command-name "gdb --annotate=1")
 '(large-file-warning-threshold nil)
 '(lsp-keymap-prefix "C-S-l")
 '(lsp-rust-all-features t)
 '(lsp-rust-analyzer-diagnostics-disabled ["unresolved-macro-call"])
 '(lsp-rust-analyzer-display-chaining-hints t)
 '(lsp-rust-analyzer-display-parameter-hints t)
 '(lsp-rust-analyzer-server-display-inlay-hints nil)
 '(lsp-rust-features [])
 '(lsp-rust-server 'rust-analyzer)
 '(lsp-signature-render-documentation nil)
 '(lsp-enable-snippet nil)
 '(lsp-ui-doc-enable nil)
 '(lsp-ui-doc-show-with-cursor nil)
 '(lsp-ui-doc-show-with-mouse nil)
 '(package-selected-packages
   '(company use-package lsp-ui eglot ## dap-mode exec-path-from-shell auctex))
 '(safe-local-variable-values
   '((TeX-master . main)
     (TeX-master . gwas-hand)
     (TeX-master . t))))


(with-eval-after-load "lsp-ui-doc"
  (setq lsp-ui-doc-enable nil)
  (setq lsp-ui-doc-show-with-cursor nil))
```

I have also checked that all the variables are actually set to nil. I am using emacs 27.2

Many thanks

**Reply by yyoncho** [reply]
Can you show me a screenshot of what you want to disable?

**Reply by shaze** [reply]
Thanks for the prompt reply

See: http://dept.ee.wits.ac.za/~scott/emacs-screenshot.png


See the minibuffer with the long explanation for what a bool is. This happens when the cursor is moved over bool (or any other keyword, type, variable etc). In this case it is a remote machine I am ssh-ing into -- so no X, GUI or mouse involved. However, I get the same behaviour on my Mac (Emacs 28.2) or if I use X.

**Reply by yyoncho** [reply]
@shaze, as you can see in bottom-right, you have started eglot as well and it is showing this message.

**Comment by shaze** [comment]
Thank you! The danger of putting thing in an init.el file that you don't understand.

**Comment by larrasket** [comment]
(setq lsp-modeline-diagnostics-enable nil)

Is it possible to make this show errors only?

**Reply by yyoncho** [reply]
@larrasket ATM there is no option for that with a setting. You can do it like this:

```
(with-eval-after-load 'lsp-modeline
  (defun lsp-modeline-diagnostics-statistics ()
    "Calculate diagnostics statistics based on `lsp-modeline-diagnostics-scope'."
    (let ((diagnostics (cond
                        ((equal :file lsp-modeline-diagnostics-scope)
                         (list (lsp--get-buffer-diagnostics)))
                        (t (->> (eq :workspace lsp-modeline-diagnostics-scope)
                                (lsp-diagnostics)
                                (ht-values)))))
          (stats (make-vector lsp/diagnostic-severity-max 0))
          strs
          (i 0))
      (mapc (lambda (buf-diags)
              (mapc (-lambda ((&Diagnostic? :severity?))
                      (when severity?
                        (cl-incf (aref stats severity?))))
                    buf-diags))
            diagnostics)
      (while (< i lsp/diagnostic-severity-max)
        (when (and (= i lsp/diagnostic-severity-error)
                   (> (aref stats i) 0))
          (setq strs
                (nconc strs
                       `(,(propertize
                           (format "%s" (aref stats i))
                           'face
                           (cond
                            ((= i lsp/diagnostic-severity-error) 'error)
                            ((= i lsp/diagnostic-severity-warning) 'warning)
                            ((= i lsp/diagnostic-severity-information) 'success)
                            ((= i lsp/diagnostic-severity-hint) 'success)))))))
        (cl-incf i))
      (-> (s-join "/" strs)
          (propertize 'mouse-face 'mode-line-highlight
                      'help-echo "mouse-1: Show diagnostics"
                      'local-map (when (require 'lsp-treemacs nil t)
                                   (make-mode-line-mouse-map
                                    'mouse-1 #'lsp-treemacs-errors-list)))))))
```

**Comment by ryukinix** [comment]
How to disable lsp open-doctor helm buffer creation when running tests? Ex.: lsp-metals. Example: https://server.manoel.dev/storage/lsp-metals-doctor-run.jpg

**Comment by philippludwig** [comment]
Man these doc popups are really a pain. I tried every setting but they still show up when the company completion menu is open, and I don't have eldoc enabled.
